warden_daemon/src/client_handler/client_command_handler.rs

```diff
@@ -0,0 +1,580 @@
+use crate::managers::realm::{Realm, RealmDescription, RealmError};
+use crate::managers::realm_configuration::RealmConfig;
+use crate::managers::warden::{Warden, WardenError};
+
+use async_trait::async_trait;
+use log::{error, info, trace};
+use serde::{Deserialize, Serialize};
+use std::io;
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
+use tokio::sync::Mutex;
+use tokio::{net::UnixStream, select};
+use tokio_util::sync::CancellationToken;
+use uuid::Uuid;
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum ClientCommand {
+    CreateRealm {
+        config: RealmConfig,
+    },
+    StartRealm {
+        uuid: Uuid,
+    },
+    StopRealm {
+        uuid: Uuid,
+    },
+    RebootRealm {
+        uuid: Uuid,
+    },
+    DestroyRealm {
+        uuid: Uuid,
+    },
+    InspectRealm {
+        uuid: Uuid,
+    },
+    ListRealms,
+    CreateApplication {
+        realm_uuid: Uuid,
+        config: ApplicationConfig,
+    },
+    StartApplication {
+        realm_uuid: Uuid,
+        application_uuid: Uuid,
+    },
+    StopApplication {
+        realm_uuid: Uuid,
+        application_uuid: Uuid,
+    },
+    UpdateApplication {
+        realm_uuid: Uuid,
+        application_uuid: Uuid,
+        config: ApplicationConfig,
+    },
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum ClientReponse {
+    Ok,
+    CreatedRealm { uuid: Uuid },
+    InspectedRealm(RealmDescription),
+    ListedRealms(Vec<RealmDescription>),
+    Error(ClientError),
+}
+
+#[derive(Debug, Error, Serialize, Deserialize, PartialEq, PartialOrd)]
+pub enum ClientError {
+    #[error("Failed to read request.")]
+    ReadingRequestFail,
+    #[error("Can't recognise a command.")]
+    UnknownCommand { length: usize },
+    #[error("Provided Uuid is invalid.")]
+    InvalidUuid,
+    #[error("Can't serialize a response.")]
+    ParsingResponseFail,
+    #[error("Warden error occured: {0}")]
+    WardenDaemonError(WardenError),
+    #[error("Realm error occured: {0}")]
+    RealmManagerError(RealmError),
+    #[error("Application error occured: {0}")]
+    ApplicationError(ApplicationError),
+    #[error("Failed to send response.")]
+    SendingResponseFail,
+}
+
+#[async_trait]
+pub trait Client {
+    async fn handle_connection(
+        warden: Arc<Mutex<Box<dyn Warden + Send + Sync>>>,
+        socket: UnixStream,
+        token: Arc<CancellationToken>,
+    ) -> Result<(), ClientError>;
+}
+
+pub struct ClientHandler {
+    warden: Arc<Mutex<Box<dyn Warden + Send + Sync>>>,
+    socket: BufReader<UnixStream>,
+    token: Arc<CancellationToken>,
+}
+
+impl ClientHandler {
+    pub async fn handle_client_requests(&mut self) -> Result<(), ClientError> {
+        loop {
+            // TODO! Refactor command reading using tokio_serde
+            let mut request_data = String::new();
+            select! {
+                readed_bytes = self.socket.read_line(&mut request_data) => {
```

Comment:
Whole communication with clients will be rewritten using tokio_serde in incoming PR. For now such solution is easier because we can socat to communicate with Warden daemon.